tools/sklearn/dict_h5_pickler.py

```diff
@@ -0,0 +1,510 @@
+        assert callable(func), "%r" % func
+
+        _args = data[_ARGS][_TUPLE]
+        args = tuple(self.load_all(_args))
+
+        try:
+            obj = args[0].__new__(args[0], * args)
+        except:
+            obj = func(*args)
+
+        _state = data.get(_STATE)
+        if _state:
+            state = self.load_all(_state)
+            setstate = getattr(obj, "__setstate__", None)
+            if setstate:
+                setstate(state)
+            else:
+                assert (type(state) is dict)
+                for k, v in state.items():
+                    setattr(obj, k, v)
+
+        self.memoize(obj)
+        return obj
+
+    def load_np_ndarray(self, data):
+        _dtype = self.load_all(data[_DTYPE])
+        _values = self.load_all(data[_VALUES])
+        obj = numpy.array(_values, dtype=_dtype)
+        return obj
+
+    def load_np_datatype(self, data):
+        _datatype = self.load_all(data[_DATATYPE])
+        _value = self.load_all(data[_VALUE])
+        obj = _datatype(_value)
+        return obj
+
+
+class H5Model:
+    """
+    Convert the dictionary model to h5 and reverse
+    """
+
+    def _create_dataset(self, file_obj, key, value):
+        """
+        Create dataset
+        """
+        try:
+            file_obj.create_dataset(key, data=json.dumps(value))
+        except:
+            file_obj.create_dataset(key, data=value)
+
+    def save_model(self, model, file_name):
+        """
+        Save the dictionary model to h5 file
+        """
+        serialised_model = ModelToDict().dump(model)
+        h5file = h5py.File(file_name, 'w')
+
+        # nested method for recursion
+        def recursive_save_model(h5file_obj, dictionary):
+            for model_key, model_value in dictionary.items():
+                type_name = type(model_value).__name__
+                try:
+                    if type_name in ['list']:
+                        if len(model_value) > 0:
+                            list_obj = all(isinstance(x, dict) for x in model_value)
+                            if list_obj is False:
+                                self._create_dataset(h5file_obj, model_key, model_value)
+                            else:
+                                for index, model_item in enumerate(model_value):
+                                    model_key_item = model_key + "/" + str(index)
+                                    if model_item is not None:
+                                        if model_key_item in h5file_obj:
+                                            recursive_save_model(model_key_item, model_item)
+                                        else:
+                                            group = h5file_obj.create_group(model_key_item)
+                                            recursive_save_model(group, model_item)
+                                    else:
+                                        self._create_dataset(h5file_obj, model_key_item, model_item)
```

Comment:
For this `else`, `model_item` is None? Sorry, I don't quite understand here.

Reply:
It takes care of complex lists containing objects of different types like `None`, `int`, `str` or `dict`. When `model_item` is `None`, just create a hierarchy with `None` as its value.